File: txt/creds.py
import base64
import json
import os
import secrets
from dataclasses import dataclass

OWNER_REQUIRED_FIELDS = [
    "owner_email",
    "cf_account_id",
    "cf_d1_database_id",
    "cf_d1_api_token",
    "display_name",
    "r2_config",
    "user_root_key",
]


@dataclass
class R2Config:
    endpoint: str
    read_write_access_key_id: str
    read_write_secret_access_key: str
    region: str
    bucket: str


@dataclass
class OwnerCreds:
    owner_email: str
    cf_account_id: str
    cf_d1_database_id: str
    cf_d1_api_token: str
    display_name: str
    r2_config: R2Config
    user_root_key: str


def load_owner_creds(path: str) -> OwnerCreds:
    data = _read_json(path)
    missing = [key for key in OWNER_REQUIRED_FIELDS if key not in data]
    if missing:
        raise ValueError(f"Missing fields in creds.json: {', '.join(missing)}")
    values = {key: data[key] for key in OWNER_REQUIRED_FIELDS if key != "r2_config"}
    values["owner_email"] = _require_owner_email(values["owner_email"])
    for field in ("cf_account_id", "cf_d1_database_id", "cf_d1_api_token"):
        values[field] = _require_nonempty(field, values[field])
    return OwnerCreds(**values, r2_config=_require_r2_config(data["r2_config"]))


def _require_r2_config(value: object) -> R2Config:
    if not isinstance(value, dict):
        raise ValueError("r2_config must be an object")
    missing = [name for name in R2Config.__annotations__ if name not in value]
    if missing:
        raise ValueError(f"Missing r2_config fields: {', '.join(missing)}")
    return R2Config(**{name: value[name] for name in R2Config.__annotations__})


def _require_owner_email(value: object) -> str:
    if not isinstance(value, str) or "@" not in value or value != value.strip():
        raise ValueError("owner_email must be a valid email address")
    return value


def _require_nonempty(name: str, value: object) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{name} must be a non-empty string")
    return value


def ensure_user_root_key(path: str, creds: OwnerCreds) -> OwnerCreds:
    if creds.user_root_key:
        return creds
    creds.user_root_key = base64.b64encode(secrets.token_bytes(256)).decode()
    data = _read_json(path)
    data["user_root_key"] = creds.user_root_key
    _write_json(path, data)
    return creds


def _read_json(path: str) -> dict:
    with open(path) as f:
        return json.load(f)


def _write_json(path: str, data: dict) -> None:
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    os.fchmod(fd, 0o600)
    with os.fdopen(fd, "w") as f:
        json.dump(data, f, indent=2)
        f.write("\n")

File: txt/creds.ts
// Loading/validating creds.json for --migrate --from-creds, identifying one
// account in a legacy sqlite snapshot. This shape is smaller than the
// browser/admin provisioning JSON: it needs the source username, wrapping
// keys, and R2 config, but not Firebase fields.
import { readFileSync } from "node:fs";
import * as C from "./constants.ts";

export interface R2ConfigResolved {
  endpoint: string;
  region: string;
  bucket: string;
  readOnlyAccessKeyId: string;
  readOnlySecretAccessKey: string;
  readWriteAccessKeyId: string | null;
  readWriteSecretAccessKey: string | null;
}

export interface Creds {
  username: string;
  usernameLookupKey: Buffer;
  userRootKey: Buffer;
  r2Config: R2ConfigResolved;
}

// Exported for reuse by other creds.json loaders (e.g. initAdminCreds.ts) --
// generic, not coupled to this file's own Creds shape.
export function requireField(value: unknown, field: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`creds.json missing/empty field: ${field}`);
  }
  return value;
}

export function checkKeyLength(
  buf: Buffer,
  minLen: number,
  field: string,
): void {
  if (buf.length < minLen) {
    throw new Error(`${field} too short (${buf.length} < ${minLen} bytes)`);
  }
}

export function loadR2Config(raw: any): R2ConfigResolved {
  const r2 = raw.r2_config ?? {};
  return {
    endpoint: requireField(r2.endpoint, "r2_config.endpoint"),
    region: requireField(r2.region, "r2_config.region"),
    bucket: requireField(r2.bucket, "r2_config.bucket"),
    readOnlyAccessKeyId: requireField(
      r2.read_only_access_key_id,
      "r2_config.read_only_access_key_id",
    ),
    readOnlySecretAccessKey: requireField(
      r2.read_only_secret_access_key,
      "r2_config.read_only_secret_access_key",
    ),
    readWriteAccessKeyId: r2.read_write_access_key_id || null,
    readWriteSecretAccessKey: r2.read_write_secret_access_key || null,
  };
}

export function hasReadWriteR2Config(r2: R2ConfigResolved): boolean {
  return Boolean(r2.readWriteAccessKeyId && r2.readWriteSecretAccessKey);
}

export function loadReadWriteR2Config(raw: any): R2ConfigResolved {
  const r2 = loadR2Config(raw);
  if (!hasReadWriteR2Config(r2)) {
    throw new Error(
      "r2_config missing read_write_access_key_id/read_write_secret_access_key",
    );
  }
  return r2;
}

// Deletion needs read-write R2 keys; --dry-run only lists, so read-only-only
// creds are tolerated there (confirmed with the user -- a deliberate
// deviation from the Python reference, which always requires read-write).
function checkWriteAccess(r2: R2ConfigResolved, dryRun: boolean): void {
  if (dryRun) return;
  if (!hasReadWriteR2Config(r2)) {
    throw new Error(
      "r2_config must include read_write_access_key_id/read_write_secret_access_key for a live (non---dry-run) run",
    );
  }
}

export function loadCreds(path: string, dryRun: boolean): Creds {
  const raw = JSON.parse(readFileSync(path, "utf8"));
  const username = requireField(raw.username, "username");
  const usernameLookupKey = Buffer.from(
    requireField(raw.username_lookup_key, "username_lookup_key"),
    "base64",
  );
  const userRootKey = Buffer.from(
    requireField(raw.user_root_key, "user_root_key"),
    "base64",
  );
  checkKeyLength(
    usernameLookupKey,
    C.USERNAME_LOOKUP_KEY_MIN_LEN,
    "username_lookup_key",
  );
  checkKeyLength(userRootKey, C.USER_ROOT_KEY_MIN_LEN, "user_root_key");
  // password is part of the shared creds shape but unused here: this tool
  // goes straight from user_root_key to umk, no login/auth step.

  const r2Config = loadR2Config(raw);
  checkWriteAccess(r2Config, dryRun);
  return { username, usernameLookupKey, userRootKey, r2Config };
}
